presto-main/src/main/java/com/facebook/presto/type/khyperloglog/KHyperLogLog.java

```diff
@@ -0,0 +1,349 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.type.khyperloglog;
+
+import com.facebook.airlift.stats.cardinality.HyperLogLog;
+import com.google.common.collect.Sets;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Murmur3Hash128;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import it.unimi.dsi.fastutil.longs.Long2DoubleMap;
+import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectRBTreeMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectSortedMap;
+import it.unimi.dsi.fastutil.longs.LongIterator;
+import it.unimi.dsi.fastutil.longs.LongRBTreeSet;
+import it.unimi.dsi.fastutil.longs.LongSortedSet;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PrimitiveIterator;
+import java.util.stream.LongStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.SIZE_OF_BYTE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.SIZE_OF_LONG;
+import static io.airlift.slice.Slices.wrappedIntArray;
+import static io.airlift.slice.Slices.wrappedLongArray;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * For reference on KHyperLogLog, see "KHyperLogLog: Estimating Reidentifiability and
+ * Joinability of Large Data at Scale" by Chia et al., 2019.
+ */
+public class KHyperLogLog
+{
+    public static final int DEFAULT_HLL_BUCKETS = 256;
+    public static final int DEFAULT_MAX_SIZE = 4096;
+    private static final byte VERSION_BYTE = 1;
+    private static final long HASH_OUTPUT_HALF_RANGE = Long.MAX_VALUE;
+    private static final int SIZE_OF_KHYPERLOGLOG = ClassLayout.parseClass(KHyperLogLog.class).instanceSize();
+    private static final int SIZE_OF_RBTREEMAP = ClassLayout.parseClass(Long2ObjectRBTreeMap.class).instanceSize();
+
+    private final Long2ObjectSortedMap<HyperLogLog> minhash;
+    private final int maxSize;
+    private final int hllBuckets;
+
+    private int hllsTotalEstimatedInMemorySize;
```

Comment:
Maybe initialize these two to 0 here instead of in the constructor?

Reply:
Ok, looks cleaner